internal/consensus/pbts_test.go

```diff
@@ -71,34 +71,45 @@ type pbtsTestConfiguration struct {
 	// The setting to use for the TimeoutPropose configuration parameter.
 	timeoutPropose time.Duration
 
-	// The timestamp of the first block produced by the network.
+	// The genesis time
 	genesisTime time.Time
 
-	// The time at which the proposal at height 2 should be delivered.
-	height2ProposalDeliverTime time.Time
+	// The timestamp of the first block produced by the network.
+	height1Time time.Time
+
+	// The times offset from height 1 block time of the block proposed at height 2.
+	height2ProposedBlockOffset time.Duration
 
-	// The timestamp of the block proposed at height 2.
-	height2ProposedBlockTime time.Time
+	// The time offset from height 1 block time at which the proposal at height 2 should be delivered.
+	height2ProposalTimeOffset time.Duration
 
-	// The timestamp of the block proposed at height 4.
-	// At height 4, the proposed block time and the deliver time are the same so
+	// The time offset from height 1 block time of the block proposed at height 4.
+	// At height 4, the proposed block and the deliver offsets are the same so
 	// that timely-ness does not affect height 4.
-	height4ProposedBlockTime time.Time
+	height4ProposedBlockOffset time.Duration
 }
 
 func newPBTSTestHarness(ctx context.Context, t *testing.T, tc pbtsTestConfiguration) pbtsTestHarness {
 	t.Helper()
 	const validators = 4
 	cfg := configSetup(t)
 	clock := new(tmtimemocks.Source)
-	if tc.height4ProposedBlockTime.IsZero() {
 
-		// Set a default height4ProposedBlockTime.
+	if tc.genesisTime.IsZero() {
+		tc.genesisTime = time.Now()
+	}
+
+	// Set this temporarily. It will be overwritten with block 1 timestamp once it is known.
+	tc.height1Time = time.Now()
+
+	if tc.height4ProposedBlockOffset == 0 {
+
+		// Set a default height4ProposedBlockOffset.
 		// Use a proposed block time that is greater than the time that the
 		// block at height 2 was delivered. Height 3 is not relevant for testing
 		// and always occurs blockTimeIota before height 4. If not otherwise specified,
 		// height 4 therefore occurs 2*blockTimeIota after height 2.
-		tc.height4ProposedBlockTime = tc.height2ProposalDeliverTime.Add(2 * blockTimeIota)
+		tc.height4ProposedBlockOffset = tc.height2ProposalTimeOffset + 20*blockTimeIota
```

Comment:
Why `20*blockTimeIota` instead of `2*blockTimeIota`? It looks like from the comment the height should occur 2 `blockTimeIota` after height2, so shouldn't we keep this as `2*`?

Reply:
good point, with `2*..` there was a consensus failure in `TestProposerWaitsForGenesisTime`:
```
2022-01-18T10:46:07+01:00 DEBUG precommit step: +2/3 prevoted proposal block; locking hash=6B82F2A5D7085F7FC0D954F96EAFF9A086486447EA3C80943BFE357B64A601D4 height=4 module=consensus round=0
2022-01-18T10:46:07+01:00 ERROR CONSENSUS FAILURE!!! err="precommit step: +2/3 prevoted for an invalid block block time 2022-01-18 09:46:07.763437 +0000 UTC not greater than last block time 2022-01-18 09:46:07.771437 +0000 UTC; relocking" module=consensus stack="goroutine 71 [running]:\nruntime/debug.Stack(0xc00055f258, 0x1953460, 0xc000eb3590)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x9f\ngithub.com/tendermint/tendermint/internal/consensus.(*State).receiveRoutine.func2(0xc000511500, 0x1c145c0)\n\t/Users/ancaz/go/src/github.com/tendermint/tendermint/internal/consensus/state.go:781 +0x5b\npanic(0x1953460, 0xc000eb3590)\n\t/usr/local/go/src/runtime/panic.go:965 +0x1b9\ngithub.com/tendermint/tendermint/internal/consensus.(*State).enterPrecommit(0xc000511500, 0x4, 0x0)\n\t/Users/ancaz/go/src/github.com/tendermint/tendermint/internal/consensus/state.go:1557 +0x1898\ngithub.com/tendermint/tendermint/internal/consensus.(*State).addVote(0xc000511500, 0xc0005d50e0, 0x0, 0x0, 0xc00055fc58, 0x8, 0xc00055fe88)\n\t/Users/ancaz/go/src/github.com/tendermint/tendermint/internal/consensus/state.go:2216 +0x13be\ngithub.com/tendermint/tendermint/internal/consensus.(*State).tryAddVote(0xc000511500, 0xc0005d50e0, 0x0, 0x0, 0xc00055fc20, 0x100f67b, 0xc000535100)\n\t/Users/ancaz/go/src/github.com/tendermint/tendermint/internal/consensus/state.go:2062 +0x56\ngithub.com/tendermint/tendermint/internal/consensus.(*State).handleMsg(0xc000511500, 0x1ce0f40, 0xc0005c2188, 0x0, 0x0)\n\t/Users/ancaz/go/src/github.com/tendermint/tendermint/internal/consensus/state.go:893 +0x96d\ngithub.com/tendermint/tendermint/internal/consensus.(*State).receiveRoutine(0xc000511500, 0x0)\n\t/Users/ancaz/go/src/github.com/tendermint/tendermint/internal/consensus/state.go:817 +0x3d2\ncreated by github.com/tendermint/tendermint/internal/consensus.(*State).startRoutines\n\t/Users/ancaz/go/src/github.com/tendermint/tendermint/internal/consensus/state.go:430 +0x14b\n"

```
But I "fixed" it in the wrong place. Just changed now to set `height4ProposedBlockOffset` in the test instead of here.